ElectricalCircuitSimulator/SimulationEngine/include/Matrix.h

```diff
@@ -2,72 +2,101 @@
 #define MATRIX_H
 
 #include <iostream>
+using std::swap;
+using std::unique_ptr;
+using std::make_unique;
 
 namespace SimulationEngine {
 
     class PLU_Factorization;
     class Matrix;
 
     // All indexes in this class are zero-based
-    class Matrix
+    class Matrix final
     {
         public:
 
             Matrix() {
                 m_iNumRows = 0;
                 m_iNumColumns = 0;
-                m_oMatrix = new double* [0];
+                m_pMatrix = new double* [0];
             };
             Matrix(const int iRows, const int iColumns);
+            Matrix(const Matrix& oMatrix);
+            Matrix(Matrix&&) = delete;
 
             ~Matrix() {
-                delete m_oMatrix;
+                delete m_pMatrix;
             }
 
-            int getNumRows() const { 
+            Matrix& operator=(const Matrix&) = delete;
+            Matrix& operator=(Matrix&&) = delete;
+
+            double& operator()(const int iRow, const int iColumn) const {
+                return m_pMatrix[iRow][iColumn];
+            }
+
+            int getNumRows() const {
                 return m_iNumRows;
             }
             int getNumColumns() const {
                 return m_iNumColumns;
             }
             double getValue(const int iRow, const int iColumn) const;
             void setValue(const int iRow, const int iColumn, const double dValue);
+            void swapRows(const int iRow1, const int iRow2) {
+                swap(m_pMatrix[iRow1], m_pMatrix[iRow2]);
+            }
+            void swapValues(const int iRow1, const int iColumn1, const int iRow2, const int iColumn2) {
+                swap(m_pMatrix[iRow1][iColumn1], m_pMatrix[iRow2][iColumn2]);
+            }
             void clear();
-            Matrix* clone() const;
             void printMatrix() const;
-            PLU_Factorization* runPLU_Factorization() const;
-            static Matrix* linearSystemSolver(const Matrix& oSourceVector, const PLU_Factorization& oPLU_Factorization);
+            static void linearSystemSolver(const Matrix& oSourceVector, const PLU_Factorization& oPLU, Matrix& oVoltageVector);
 
         private:
 
             int m_iNumRows;
             int m_iNumColumns;
-            double** m_oMatrix; // Outer pointer is rows, inner pointer is columns
+            double** m_pMatrix; // Outer pointer is rows, inner pointer is columns
```

Comment:
You can avoid doing the rule of 5 thing entirely by using smart pointers here instead of raw arrays

`std::unique_ptr< std::unique_ptr< double[] >[] > _matrix`

(I'd argue to drop the hungarian notation too, it's so unhelpful: modern IDEs will tell you the type, and we're going to try to write code so that it might be robust and types could be easily changed if need be--using auto anywhere we can--and then we don't want to have to rename variables)

in the constructor, you can initialize it as
`_matrix(make_unique<unique_ptr<double[]>[]>(_numRows))`

and then in the body you can do the for loop to initialize each row
`for(auto& row : _matrix)`
`  row = make_unique<double[]>(_numColumns);`

Reply:
Will do later in Matrix rework